subsys/zigbee/common/zigbee_ota_client.c

```diff
@@ -0,0 +1,564 @@
+
+/* OTA cluster attributes data */
+ZB_ZCL_DECLARE_OTA_UPGRADE_ATTRIB_LIST(ota_upgrade_attr_list,
+				dev_ctx.ota_attr.upgrade_server,
+				&dev_ctx.ota_attr.file_offset,
+				&dev_ctx.ota_attr.file_version,
+				&dev_ctx.ota_attr.stack_version,
+				&dev_ctx.ota_attr.downloaded_file_ver,
+				&dev_ctx.ota_attr.downloaded_stack_ver,
+				&dev_ctx.ota_attr.image_status,
+				&dev_ctx.ota_attr.manufacturer,
+				&dev_ctx.ota_attr.image_type,
+				&dev_ctx.ota_attr.min_block_reque,
+				&dev_ctx.ota_attr.image_stamp,
+				&dev_ctx.ota_attr.server_addr,
+				&dev_ctx.ota_attr.server_ep,
+				(uint16_t)CONFIG_ZIGBEE_OTA_HW_VERSION,
+				CONFIG_ZIGBEE_OTA_DATA_BLOCK_SIZE,
+				ZB_ZCL_OTA_UPGRADE_QUERY_TIMER_COUNT_DEF);
+
+ZB_HA_DECLARE_OTA_UPGRADE_CLIENT_CLUSTER_LIST(ota_upgrade_client_clusters,
+	ota_basic_attr_list, ota_upgrade_attr_list);
+
+ZB_HA_DECLARE_OTA_UPGRADE_CLIENT_EP(ota_upgrade_client_ep,
+				    CONFIG_ZIGBEE_OTA_ENDPOINT,
+				    ota_upgrade_client_clusters);
+
+
+/**@brief Function for initializing all OTA clusters attributes.
+ *
+ * @note This function shall be called after the dfu initialization.
+ */
+static void ota_client_attr_init(void)
+{
+	struct mcuboot_img_header mcuboot_header;
+	union app_ver ota_app_ver;
+
+	int err = boot_read_bank_header(PM_APP_ID,
+					&mcuboot_header,
+					sizeof(mcuboot_header));
+
+	if (err) {
+		LOG_WRN("Failed read app ver from image header (err %d)", err);
+		ota_app_ver.uint32_ver = 0;
+	} else {
+		ota_app_ver.major = mcuboot_header.h.v1.sem_ver.major;
+		ota_app_ver.minor = mcuboot_header.h.v1.sem_ver.minor;
+		ota_app_ver.revision =  mcuboot_header.h.v1.sem_ver.revision;
+	}
+
+
+	/* Basic cluster attributes data */
+	dev_ctx.basic_attr.zcl_version  = ZB_ZCL_VERSION;
+	dev_ctx.basic_attr.power_source = ZB_ZCL_BASIC_POWER_SOURCE_UNKNOWN;
+
+	/* OTA cluster attributes data */
+	zb_ieee_addr_t addr = ZB_ZCL_OTA_UPGRADE_SERVER_DEF_VALUE;
+
+	ZB_MEMCPY(dev_ctx.ota_attr.upgrade_server, addr,
+		  sizeof(zb_ieee_addr_t));
+	dev_ctx.ota_attr.file_offset = ZB_ZCL_OTA_UPGRADE_FILE_OFFSET_DEF_VALUE;
+	dev_ctx.ota_attr.file_version = ota_app_ver.uint32_ver;
+	dev_ctx.ota_attr.stack_version =
+		ZB_ZCL_OTA_UPGRADE_FILE_HEADER_STACK_PRO;
+	dev_ctx.ota_attr.downloaded_file_ver  =
+		ZB_ZCL_OTA_UPGRADE_DOWNLOADED_FILE_VERSION_DEF_VALUE;
+	dev_ctx.ota_attr.downloaded_stack_ver =
+		ZB_ZCL_OTA_UPGRADE_DOWNLOADED_STACK_DEF_VALUE;
+	dev_ctx.ota_attr.image_status =
+		ZB_ZCL_OTA_UPGRADE_IMAGE_STATUS_DEF_VALUE;
+	dev_ctx.ota_attr.manufacturer = CONFIG_ZIGBEE_OTA_MANUFACTURER_ID;
+	dev_ctx.ota_attr.image_type = CONFIG_ZIGBEE_OTA_IMAGE_TYPE;
+	dev_ctx.ota_attr.min_block_reque = 0;
+	dev_ctx.ota_attr.image_stamp = ZB_ZCL_OTA_UPGRADE_IMAGE_STAMP_MIN_VALUE;
+}
+
+
+static void zb_ota_clt_upinit(void)
+{
+	/* Processing optional header switched off by default */
+	ota_ctx.optional_header_finished = true;
+}
+
+
+/** @brief Function to abort the firmware upgrade process
+ */
+static void zb_abort_dfu(void)
```

Comment:
I sense some inconsistency in naming convention. There are many prefixes/suffixes (`ota`, `client`, `zb_ota`, `ota_clt`, `dfu`), so it may be convenient for a reader if we decide to use a bit smaller subset of them.